lib/minitest/reporters/ruby_lsp_reporter.rb

```diff
@@ -0,0 +1,96 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "ruby_lsp/test_reporting"
+
+module Minitest
+  module Reporters
+    class RubyLspReporter < ::Minitest::Reporters::BaseReporter
+      extend T::Sig
+
+      sig { void }
+      def initialize
+        @reporting = T.let(RubyLsp::TestReporter.new, RubyLsp::TestReporter)
+        super
+      end
+
+      sig { params(test: Minitest::Test).void }
+      def before_test(test)
+        @reporting.before_test(
+          id: id_from_test(test),
+          file: file_for_class_name(test),
+        )
+        super
+      end
+
+      sig { params(test: Minitest::Test).void }
+      def after_test(test)
+        @reporting.after_test(
+          id: id_from_test(test),
+          file: file_for_class_name(test),
+        )
+        super
+      end
+
+      sig { params(test: Minitest::Result).void }
+      def record(test)
+        super
+
+        if test.passed?
+          record_pass(test)
+        elsif test.skipped?
+          record_skip(test)
+        elsif test.failure
+          record_fail(test)
+        end
+      end
+
+      sig { params(result: Minitest::Result).void }
+      def record_pass(result)
+        info = {
+          id: id_from_result(result),
+          file: result.source_location[0],
+        }
+        @reporting.record_pass(**info)
+      end
+
+      sig { params(result: Minitest::Result).void }
+      def record_skip(result)
+        info = {
+          id: id_from_result(result),
+          message: result.failure.message,
+          file: result.source_location[0],
+        }
+        @reporting.record_skip(**info)
+      end
+
+      sig { params(result: Minitest::Result).void }
+      def record_fail(result)
+        info = {
+          id: id_from_result(result),
+          type: result.failure.class.name,
+          message: result.failure.message,
+          file: result.source_location[0],
+        }
+        @reporting.record_fail(**info)
+      end
+
+      private
+
+      sig { params(test: Minitest::Test).returns(String) }
+      def id_from_test(test)
+        [test.class.name, test.name].join("#")
+      end
+
+      sig { params(result: Minitest::Result).returns(String) }
+      def id_from_result(result)
+        [result.name, result.klass].join("#")
+      end
+
+      sig { params(test: Minitest::Test).returns(String) }
+      def file_for_class_name(test)
+        T.must(Kernel.const_source_location(test.class_name)).first
```

Comment:
Just a headsup that even in minitest/test-unit it's possible to define test classes dynamically, which make their source location path a bit different, like:

```rb
class Foo
  class_eval <<~RUBY
    class FooTest < Test::Unit::TestCase
      def test_foo
        puts "========================================="
        puts Kernel.const_source_location("Foo::FooTest")
        puts "========================================="
        fail
      end
    end
  RUBY
end
```

```
=========================================
(eval at /Users/hung-wulo/src/github.com/Shopify/rdoc/test/rdoc/test_rdoc_comment.rb:483)
1
=========================================
```

Reply:
Pushed an update to handle that. Still need to think about what we do in the UI.